nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/certificate/CertificateRealm.java

```diff
@@ -175,6 +184,71 @@ public String authenticate(Subject subject, X500Principal principal) {
         return principalName;
     }
 
+    private void validateSubjectViaAPI(Subject subject, X500Principal principal) {
+        X509Certificate certificate = getCertificateFromSubject(subject, principal);
+        if (certificate == null) {
+            _logger.warning(() -> String.format("No X509Certificate found(subject=%s, principal=%s)", subject, principal));
+            return;
+        }
+
+        // Ask all the Client Certificate Validator.
+        List<ClientCertificateValidator> validators = Collections.emptyList();
+        try {
+            validators = loadValidatorClasses();
+        } catch (Throwable exc) {
+            _logger.log(Level.WARNING, "Exception while loading certificate validation class", exc);
+            clientCertificateValidatorMap.remove(Utility.getClassLoader());
+        }
+
+        boolean failed = false;
+        if (!validators.isEmpty()) {
+            for (ClientCertificateValidator validator : validators) {
+                if (!validator.isValid(subject, principal, certificate)) {
+                    _logger.info(() -> String.format("Client Certificate validation failed for (subject=%s, principal=%s) by %s"
```

Comment:
Is this worth a warning instead of an info log?

Reply:
I don't think so. There is also not a Warning when an invalid username/password is presented.